src/main/java/com/endava/internship/service/UserServiceImpl.java

```diff
@@ -4,66 +4,99 @@
 import com.endava.internship.domain.User;
 import com.endava.internship.service.UserService;
 
-import java.util.List;
-import java.util.Map;
-import java.util.Optional;
+import java.util.*;
 import java.util.function.Function;
 import java.util.function.Predicate;
+import java.util.stream.Collectors;
 
 public class UserServiceImpl implements UserService {
 
     @Override
     public List<String> getFirstNamesReverseSorted(List<User> users) {
-        throw new UnsupportedOperationException("Not implemented");
+        return users.stream().
+                map(User::getFirstName).
+                sorted(Comparator.reverseOrder()).
+                collect(Collectors.toList());
     }
 
     @Override
     public List<User> sortByAgeDescAndNameAsc(final List<User> users) {
-        throw new UnsupportedOperationException("Not implemented");
+        return users.stream().
+                sorted(Comparator.comparing(User::getAge).
+                        reversed().
+                        thenComparing(User::getFirstName)).
+                collect(Collectors.toList());
     }
 
     @Override
     public List<Privilege> getAllDistinctPrivileges(final List<User> users) {
-        throw new UnsupportedOperationException("Not implemented");
+        return users.stream().
+                flatMap(user -> user.getPrivileges().stream()).
+                collect(Collectors.toList());
     }
 
     @Override
     public Optional<User> getUpdateUserWithAgeHigherThan(final List<User> users, final int age) {
-        throw new UnsupportedOperationException("Not implemented");
+        return users.stream().
+                filter(user -> user.getAge() > age || user.getPrivileges().
```

Comment:
user.getPrivileges() returns a list, list can't be compared with Privilege using equals

Reply:
use list method `contains`

Reply:
has been changed